| **Inicio** | **atrás 14** | **Siguiente 16** |
|----------- |-------------- |---------------|
| [🏠](../../README.md) | [⏪](./14.Scripts_Modulos.ipynb)| [⏩](./16.Modulos_Numpy.ipynb)|

# **15. Módulos de Python: math y cmath**

## **El módulo math**

El módulo `math` es un módulo integrado de Python que proporciona funciones matemáticas y constantes. Este módulo contiene una amplia variedad de funciones matemáticas que pueden ser utilizadas para realizar operaciones aritméticas y cálculos matemáticos más complejos.

Aquí hay algunos ejemplos de cómo se puede utilizar el módulo `math` en Python:

In [1]:
import math

# Valor absoluto
print(abs(-5))  # 5

# Redondear números
print(round(3.14159, 2))  # 3.14

# Constantes
print(math.pi)  # 3.141592653589793
print(math.e)  # 2.718281828459045

# Funciones trigonométricas
print(math.sin(math.pi/2))  # 1.0
print(math.cos(math.pi/4))  # 0.7071067811865476

# Funciones exponenciales y logarítmicas
print(math.exp(2))  # 7.3890560989306495
print(math.log(10))  # 2.302585092994046


5
3.14
3.141592653589793
2.718281828459045
1.0
0.7071067811865476
7.38905609893065
2.302585092994046


En este ejemplo, se importa el módulo `math` y se utilizan algunas de sus funciones y constantes. La función `abs()` se utiliza para calcular el valor absoluto de un número, mientras que la función `round()` se utiliza para redondear un número a un número específico de decimales.

El módulo `math` también proporciona algunas constantes matemáticas comunes, como `pi` y `e`. Estas constantes se pueden utilizar en cálculos matemáticos.

Las funciones trigonométricas como `sin()` y `cos()` se pueden utilizar para calcular el seno y el coseno de un ángulo dado. Las funciones exponenciales y logarítmicas como `exp()` y `log()` se pueden utilizar para calcular el valor exponencial y el logaritmo natural de un número, respectivamente.

En resumen, el módulo `math` es una herramienta útil para realizar cálculos matemáticos en Python. Ofrece una variedad de funciones y constantes matemáticas que pueden ser utilizadas para realizar cálculos precisos y eficientes.

## **Constantes del módulo math**

El módulo `math` de Python proporciona varias constantes matemáticas que se pueden utilizar en cálculos matemáticos. Aquí hay algunos ejemplos de constantes comunes del módulo `math`:

* `math.pi`: la constante pi (π) es la relación entre la circunferencia y el diámetro de un círculo. Su valor aproximado es 3.141592653589793.

In [2]:
import math

print(math.pi)  # 3.141592653589793


3.141592653589793


* `math.e`: la constante e es una constante matemática importante que es la base de los logaritmos naturales. Su valor aproximado es 2.718281828459045.

In [3]:
import math

print(math.e)  # 2.718281828459045


2.718281828459045


* `math.tau`: la constante tau (τ) es otra constante matemática que es igual a 2π. Es útil para algunos cálculos matemáticos, especialmente en geometría. Su valor aproximado es 6.283185307179586.

In [4]:
import math

print(math.tau)  # 6.283185307179586


6.283185307179586


* `math.inf`: esta constante representa el infinito matemático positivo. Puede ser utilizado en cálculos matemáticos para representar valores que son mayores que cualquier número finito.

In [5]:
import math

x = math.inf
print(x)  # inf
print(x + 1)  # inf
print(x * 2)  # inf


inf
inf
inf


* `math.nan`: esta constante representa "no un número" (NaN), que es un valor especial que se utiliza en cálculos matemáticos para representar resultados indeterminados o imposibles.

In [6]:
import math

x = math.nan
print(x)  # nan
print(x + 1)  # nan
print(x * 2)  # nan


nan
nan
nan


Estas son solo algunas de las constantes que se pueden encontrar en el módulo `math`. Además de estas constantes, el módulo `math` también proporciona funciones matemáticas para realizar cálculos más complejos, como funciones trigonométricas, logarítmicas, exponenciales y más.

## **Métodos. Parte entera de un número**

En Python, el método `int()` se puede utilizar para obtener la parte entera de un número. Este método convierte un número en un entero, eliminando cualquier parte decimal. Si el número es negativo, la parte entera se redondea hacia cero.

**Ejemplo:**

In [7]:
num = 3.14159
integer = int(num)
print(integer)  # Output: 3

num2 = -4.8
integer2 = int(num2)
print(integer2)  # Output: -4


3
-4


También podemos usar el método `math.floor()` del módulo `math` para obtener la parte entera de un número, que redondea el número a la parte entera inferior.

**Ejemplo:**

In [8]:
import math

num = 3.14159
integer = math.floor(num)
print(integer)  # Output: 3

num2 = -4.8
integer2 = math.floor(num2)
print(integer2)  # Output: -5


3
-5


Además, también se puede usar el método `math.trunc()` para obtener la parte entera de un número, que simplemente elimina la parte decimal del número sin redondear.

**Ejemplo:**

In [9]:
import math

num = 3.14159
integer = math.trunc(num)
print(integer)  # Output: 3

num2 = -4.8
integer2 = math.trunc(num2)
print(integer2)  # Output: -4


3
-4


En resumen, en Python hay varios métodos que se pueden utilizar para obtener la parte entera de un número, como `int()`, `math.floor()` y `math.trunc()`, cada uno con sus propias características y uso dependiendo del caso específico.

## **Métodos. Aritmética**

En Python, hay varios métodos predefinidos para realizar operaciones aritméticas con números. A continuación, se explican algunos de los más comunes:

* **+ :** suma dos números
* **- :** resta dos números
* `*` : multiplica dos números
* **/ :** divide dos números y devuelve el resultado como un flotante
* **// :** divide dos números y devuelve el resultado como un entero (parte entera)
* **% :** divide dos números y devuelve el resto
* `**` : eleva un número a una potencia

**Ejemplos:**

In [10]:
a = 5
b = 2

# Suma
print(a + b)  # Output: 7

# Resta
print(a - b)  # Output: 3

# Multiplicación
print(a * b)  # Output: 10

# División
print(a / b)  # Output: 2.5

# División entera
print(a // b)  # Output: 2

# Resto
print(a % b)  # Output: 1

# Potencia
print(a ** b)  # Output: 25


7
3
10
2.5
2
1
25


Además de estos operadores, también existen algunas funciones matemáticas predefinidas en el módulo `math`. Por ejemplo:

In [11]:
import math

a = 5
b = 2

# Raíz cuadrada
print(math.sqrt(a))  # Output: 2.23606797749979

# Logaritmo natural
print(math.log(a))  # Output: 1.6094379124341003

# Logaritmo base 10
print(math.log10(a))  # Output: 0.6989700043360189

# Seno
print(math.sin(a))  # Output: -0.9589242746631385

# Coseno
print(math.cos(a))  # Output: 0.28366218546322625


2.23606797749979
1.6094379124341003
0.6989700043360189
-0.9589242746631385
0.28366218546322625


En resumen, Python cuenta con una amplia variedad de métodos y funciones matemáticas para realizar operaciones aritméticas y cálculos más complejos. Los operadores básicos `+`, `-`, `*`, `/`, `//`, `%`, y `**` son fundamentales para realizar operaciones sencillas, mientras que las funciones del módulo `math` permiten realizar operaciones más complejas como cálculos trigonométricos, logarítmicos, exponenciales, entre otros.

## **Métodos. Potencias y logaritmos**

En Python, el módulo `math` proporciona una variedad de métodos para calcular potencias y logaritmos de diferentes bases. A continuación, se detallan algunos de estos métodos con ejemplos:

**1. `math.pow(x, y)`:**

 Este método devuelve el valor de `x` elevado a la potencia `y`. Ambos argumentos pueden ser enteros o flotantes. Ejemplo:

In [12]:
import math

print(math.pow(2, 3)) # Salida: 8.0
print(math.pow(3, 2)) # Salida: 9.0


8.0
9.0


**2. `math.sqrt(x)`:**

Este método devuelve la raíz cuadrada de `x`. Ejemplo:

In [13]:
import math

print(math.sqrt(16)) # Salida: 4.0


4.0


**3. `math.exp(x)`:**

 Este método devuelve `e` elevado a la potencia `x`, donde `e` es la constante matemática `e` (aproximadamente 2.71828). Ejemplo:

In [14]:
import math

print(math.exp(2)) # Salida: 7.3890560989306495


7.38905609893065


**4. `math.log(x[, base])`:**

 Este método devuelve el logaritmo natural de `x` (si se omite el argumento opcional base) o el logaritmo de `x` en la base dada. Ejemplo:

In [15]:
import math

print(math.log(10)) # Salida: 2.302585092994046
print(math.log(10, 2)) # Salida: 3.3219280948873626


2.302585092994046
3.3219280948873626


**5. `math.log10(x)`:**

 Este método devuelve el logaritmo de `x` en la base `10`. Ejemplo:

In [16]:
import math

print(math.log10(100)) # Salida: 2.0


2.0


**6. `math.log2(x)`:**

 Este método devuelve el logaritmo de `x` en la base `2`. Ejemplo:

In [17]:
import math

print(math.log2(16)) # Salida: 4.0


4.0


En resumen, los métodos en el módulo `math` permiten realizar cálculos matemáticos avanzados en Python, incluyendo potencias y logaritmos de diferentes bases.

## **Métodos. MCD y MCM**

En Python, se pueden utilizar los métodos `gcd()` y `lcm()` del módulo `math` para encontrar el máximo común divisor `(MCD)` y el mínimo común múltiplo `(MCM)` de dos o más números enteros.

El método `gcd()` devuelve el `MCD` de dos o más números enteros. Por ejemplo:

In [18]:
import math

num1 = 24
num2 = 36

mcd = math.gcd(num1, num2)

print("El MCD de", num1, "y", num2, "es:", mcd)


El MCD de 24 y 36 es: 12


En este ejemplo, se importa el módulo `math` y se definen dos números enteros `num1` y `num2`. Luego, se utiliza el método `gcd()` para encontrar el MCD de ambos números y se guarda en la variable `mcd`. Finalmente, se imprime el resultado.

El método `lcm()` devuelve el MCM de dos o más números enteros. Por ejemplo:

In [19]:
import math

num1 = 3
num2 = 5

mcm = math.lcm(num1, num2)

print("El MCM de", num1, "y", num2, "es:", mcm)


El MCM de 3 y 5 es: 15


En este ejemplo, se importa el módulo `math` y se definen dos números enteros `num1` y `num2`. Luego, se utiliza el método `lcm()` para encontrar el MCM de ambos números y se guarda en la variable `mcm`. Finalmente, se imprime el resultado.

Es importante tener en cuenta que el método `lcm()` solo está disponible a partir de la versión 3.9 de Python. Si se utiliza una versión anterior, se puede utilizar la siguiente función para encontrar el MCM de dos o más números enteros:

In [20]:
def lcm(numbers):
    lcm = 1
    for number in numbers:
        lcm = lcm * number // math.gcd(lcm, number)
    return lcm


Esta función toma una lista de números enteros como argumento y utiliza un bucle `for` para calcular el `MCM` utilizando la fórmula:

`MCM = (num1 * num2) / MCD(num1, num2)`

donde `num1` y `num2` son dos números de la lista y `MCD()` es el método `gcd()` del módulo `math`. La función devuelve el `MCM` de todos los números en la lista.

Por ejemplo:

In [22]:
import math

def lcm(numbers):
    lcm = 1
    for number in numbers:
        lcm = lcm * number // math.gcd(lcm, number)
    return lcm

numeros = [2, 4, 6, 8, 10]

mcm = lcm(numeros)

print("El MCM de", numeros, "es:", mcm)


El MCM de [2, 4, 6, 8, 10] es: 120


En este ejemplo, se importa el módulo `math` y se define una función `lcm()` que toma una lista de números enteros como argumento y devuelve el `MCM` de todos los números en la lista. Luego, se define una lista de números enteros numeros y se utiliza la función `lcm()` para encontrar el `MCM` de todos los números en la lista.

## **Métodos. Combinatoria**

En Python, el módulo `math` incluye métodos para realizar cálculos combinatorios, que son operaciones matemáticas relacionadas con la elección y disposición de elementos de un conjunto. Estos métodos son útiles en diversas aplicaciones, como la estadística, la teoría de la probabilidad y la teoría de juegos, entre otros.

A continuación, se describen algunos de los métodos de combinatoria más comunes disponibles en el módulo `math`:

* **factorial(n):**

 Devuelve el factorial de un número entero `n`. El factorial de un número es el producto de todos los números enteros positivos desde 1 hasta `n`. Por ejemplo, el factorial de 5 es 5 x 4 x 3 x 2 x 1 = 120. Ejemplo:

In [23]:
import math

n = 5
factorial_n = math.factorial(n)
print("Factorial de", n, "es", factorial_n)


Factorial de 5 es 120


* **comb(n, k):**

 Devuelve el número de combinaciones de `n` elementos tomados de `k` en `k`, sin repetición y sin importar el orden. Por ejemplo, si se tienen 5 elementos y se quieren tomar 3, el número de combinaciones posibles es 5! / (3! x (5 - 3)!) = 10. Ejemplo:

In [24]:
import math

n = 5
k = 3
num_comb = math.comb(n, k)
print("Número de combinaciones de", n, "tomados de", k, "es", num_comb)


Número de combinaciones de 5 tomados de 3 es 10


* **perm(n, k):**

 Devuelve el número de permutaciones de `n` elementos tomados de `k` en `k`, sin repetición y teniendo en cuenta el orden. Por ejemplo, si se tienen 5 elementos y se quieren tomar 3, el número de permutaciones posibles es 5! / (5 - 3)! = 60. Ejemplo:

In [25]:
import math

n = 5
k = 3
num_perm = math.perm(n, k)
print("Número de permutaciones de", n, "tomados de", k, "es", num_perm)


Número de permutaciones de 5 tomados de 3 es 60


* **gcd(a, b):**

 Devuelve el máximo común divisor de dos números enteros `a` y `b`. Ejemplo:

In [26]:
import math

a = 12
b = 8
mcd_ab = math.gcd(a, b)
print("El máximo común divisor de", a, "y", b, "es", mcd_ab)


El máximo común divisor de 12 y 8 es 4


* **isclose(a, b, rel_tol=1e-9, abs_tol=0.0):**

 Compara dos números `a` y `b` para determinar si son `"cercanos"` en términos relativos o absolutos, utilizando tolerancias relativas y absolutas dadas. Este método es útil cuando se trabaja con números en punto flotante, ya que la comparación directa de estos números puede llevar a errores debido a la imprecisión de la representación en punto flotante. Ejemplo:

In [27]:
import math

a = 0.1 + 0.1 + 0.1
b = 0.3
son_cercanos = math.isclose(a, b)
print("¿", a, "y", b, "son cercanos?", son_cercanos)


¿ 0.30000000000000004 y 0.3 son cercanos? True


## **Métodos. Valor absoluto y signo**

En Python, los métodos para calcular el valor absoluto y el signo de un número se encuentran en el módulo `math`. A continuación, se explica cada uno de estos métodos con ejemplos:

* **fabs(x):**

 Este método devuelve el valor absoluto de `x`. La función toma un número real como argumento y devuelve su valor absoluto.
Ejemplo:

In [28]:
import math

x = -4.5
y = 3

print("Valor absoluto de x: ", math.fabs(x))
print("Valor absoluto de y: ", math.fabs(y))


Valor absoluto de x:  4.5
Valor absoluto de y:  3.0


* **copysign(x, y):**

 Este método devuelve un número con el valor absoluto de `x` y el signo de `y`. La función toma dos números como argumentos y devuelve un número con el valor absoluto de `x` y el signo de `y`.
Ejemplo:

In [29]:
import math

x = -4.5
y = 3

print("Número con valor absoluto de x y signo de y: ", math.copysign(x, y))
print("Número con valor absoluto de y y signo de x: ", math.copysign(y, x))


Número con valor absoluto de x y signo de y:  4.5
Número con valor absoluto de y y signo de x:  -3.0


En resumen, el método `fabs` devuelve el valor absoluto de un número, mientras que el método `copysign` devuelve un número con el valor absoluto de un número y el signo de otro número.

## **Métodos. Funciones trigonométricas**

En Python, el módulo `math` proporciona un conjunto de funciones trigonométricas que se pueden utilizar para realizar cálculos trigonométricos. Estas funciones se pueden utilizar para calcular senos, cosenos, tangentes, cotangentes, arcos, etc.

A continuación, se presentan algunas de las funciones trigonométricas disponibles en el módulo `math`:

* **math.sin(x):** Devuelve el seno de x (en radianes).
* **math.cos(x):** Devuelve el coseno de x (en radianes).
* **math.tan(x):** Devuelve la tangente de x (en radianes).
* **math.asin(x):** Devuelve el arco seno de x (en radianes).
* **math.acos(x):** Devuelve el arco coseno de x (en radianes).
* **math.atan(x):** Devuelve el arco tangente de x (en radianes).

Cabe destacar que estas funciones trabajan con ángulos en radianes, no en grados. Si se tiene un ángulo en grados, es necesario convertirlo a radianes antes de utilizar estas funciones.

A continuación se presentan algunos ejemplos de cómo utilizar estas funciones:

In [30]:
import math

# Calcular el seno, coseno y tangente de un ángulo
angulo = math.pi/4
seno = math.sin(angulo)
coseno = math.cos(angulo)
tangente = math.tan(angulo)
print("Ángulo: ", angulo)
print("Seno: ", seno)
print("Coseno: ", coseno)
print("Tangente: ", tangente)

# Calcular el arco seno, coseno y tangente de un valor
valor = 0.5
arco_seno = math.asin(valor)
arco_coseno = math.acos(valor)
arco_tangente = math.atan(valor)
print("Valor: ", valor)
print("Arco seno: ", arco_seno)
print("Arco coseno: ", arco_coseno)
print("Arco tangente: ", arco_tangente)


Ángulo:  0.7853981633974483
Seno:  0.7071067811865475
Coseno:  0.7071067811865476
Tangente:  0.9999999999999999
Valor:  0.5
Arco seno:  0.5235987755982989
Arco coseno:  1.0471975511965979
Arco tangente:  0.4636476090008061


En este ejemplo, se importa el módulo `math` y se utilizan varias de sus funciones trigonométricas para calcular el seno, coseno, tangente, arco seno, arco coseno y arco tangente de un ángulo y de un valor dado. Cada función devuelve el resultado correspondiente que se imprime en la consola.

## **Métodos. Funciones hiperbólicas**

En Python, el módulo `math` también incluye métodos para calcular funciones hiperbólicas, tales como seno hiperbólico, coseno hiperbólico, tangente hiperbólica, entre otros.

A continuación se presentan algunos ejemplos de cómo utilizar estas funciones:

* El método `math.sinh(x)` devuelve el seno hiperbólico de `x`.

In [31]:
import math

x = 1.5
seno_hiperbolico = math.sinh(x)
print(seno_hiperbolico) # Output: 2.1292794550948173


2.1292794550948173


* El método `math.cosh(x)` devuelve el coseno hiperbólico de `x`.

In [32]:
import math

x = 1.5
coseno_hiperbolico = math.cosh(x)
print(coseno_hiperbolico) # Output: 2.3524096152432476


2.352409615243247


* El método `math.tanh(x)` devuelve la tangente hiperbólica de `x`.

In [33]:
import math

x = 1.5
tangente_hiperbolica = math.tanh(x)
print(tangente_hiperbolica) # Output: 0.9051482536448666


0.9051482536448664


* El método `math.asinh(x)` devuelve el arcoseno hiperbólico de `x`.

In [34]:
import math

x = 1.5
arcoseno_hiperbolico = math.asinh(x)
print(arcoseno_hiperbolico) # Output: 1.1947632172871093


1.1947632172871094


* El método `math.acosh(x)` devuelve el arcocoseno hiperbólico de `x`.

In [35]:
import math

x = 1.5
arcocoseno_hiperbolico = math.acosh(x)
print(arcocoseno_hiperbolico) # Output: 0.9624236501192069


0.9624236501192069


* El método `math.atanh(x)` devuelve el arcotangente hiperbólica de `x`.

In [36]:
import math

x = 0.8
arcotangente_hiperbolica = math.atanh(x)
print(arcotangente_hiperbolica) # Output: 1.0986122886681098


1.0986122886681098


## **Métodos. Funciones de clasificación**

En Python, los métodos y funciones de clasificación son herramientas útiles para organizar y analizar datos.

Los métodos son funciones que están asociadas a un objeto particular y se utilizan para realizar operaciones específicas en ese objeto. Por ejemplo, el método `sort()` se utiliza para ordenar los elementos de una lista en Python.

Por otro lado, las funciones de clasificación son funciones que se utilizan para clasificar o agrupar elementos de datos en función de ciertos criterios. Estos criterios pueden ser valores numéricos, cadenas de caracteres o cualquier otro tipo de datos.

A continuación, te presento algunos ejemplos de métodos y funciones de clasificación en Python:

* **Método sort():**

Este método se utiliza para ordenar los elementos de una lista en orden ascendente o descendente. Por ejemplo:

In [38]:
lista = [5, 2, 9, 1, 7]
lista.sort() # Ordenar la lista en orden ascendente
print(lista) # Salida: [1, 2, 5, 7, 9]


[1, 2, 5, 7, 9]


* **Método reverse():**

Este método se utiliza para invertir el orden de los elementos de una lista. Por ejemplo:

In [1]:
lista = [1, 2, 3, 4, 5]
lista.reverse() # Invertir el orden de la lista
print(lista) # Salida: [5, 4, 3, 2, 1]


[5, 4, 3, 2, 1]


* **Función sorted():**

Esta función se utiliza para ordenar una lista sin modificar la lista original. Por ejemplo:

In [2]:
lista = [5, 2, 9, 1, 7]
ordenada = sorted(lista) # Ordenar la lista sin modificar la original
print(ordenada) # Salida: [1, 2, 5, 7, 9]


[1, 2, 5, 7, 9]


* **Función filter():**

Esta función se utiliza para filtrar elementos de una lista según un criterio específico. Por ejemplo, si queremos filtrar solo los números pares de una lista, podemos hacer lo siguiente:

In [3]:
lista = [1, 2, 3, 4, 5, 6, 7, 8, 9]
pares = list(filter(lambda x: x % 2 == 0, lista)) # Filtrar solo los números pares
print(pares) # Salida: [2, 4, 6, 8]


[2, 4, 6, 8]


* **Función map():**

Esta función se utiliza para aplicar una función a cada elemento de una lista. Por ejemplo, si queremos calcular el cuadrado de cada número en una lista, podemos hacer lo siguiente:

In [4]:
lista = [1, 2, 3, 4, 5]
cuadrados = list(map(lambda x: x ** 2, lista)) # Calcular el cuadrado de cada número en la lista
print(cuadrados) # Salida: [1, 4, 9, 16, 25]


[1, 4, 9, 16, 25]


## **Métodos. Funciones especiales**

En Python, los métodos y funciones especiales, también conocidas como `"métodos mágicos"`, son herramientas que permiten personalizar el comportamiento de las clases y objetos en el lenguaje.

Los métodos especiales se identifican por tener un doble guión bajo antes y después de su nombre, por ejemplo `__init__()` es el método especial utilizado para inicializar un objeto de una clase. Estos métodos permiten definir comportamientos específicos para operaciones como la comparación de objetos, la conversión a cadenas de caracteres, entre otras.

A continuación, te presento algunos ejemplos de métodos especiales en Python:

* **Método `__init__()`:**

Este método se utiliza para inicializar un objeto de una clase. Es llamado automáticamente cuando se crea una nueva instancia de la clase. Por ejemplo:

In [5]:
class Persona:
    def __init__(self, nombre, edad):
        self.nombre = nombre
        self.edad = edad

p = Persona("Juan", 25) # Crear un objeto de la clase Persona


* **Método `__str__()`:**

Este método se utiliza para devolver una representación en cadena de caracteres de un objeto. Es llamado automáticamente cuando se llama a la función `str()` en un objeto. Por ejemplo:

In [6]:
class Persona:
    def __init__(self, nombre, edad):
        self.nombre = nombre
        self.edad = edad

    def __str__(self):
        return f"{self.nombre} tiene {self.edad} años."

p = Persona("Juan", 25) # Crear un objeto de la clase Persona
print(str(p)) # Salida: "Juan tiene 25 años."


Juan tiene 25 años.


* **Método `__eq__()`:**

Este método se utiliza para comparar si dos objetos son iguales. Es llamado automáticamente cuando se usa el operador `==` en dos objetos. Por ejemplo:

In [7]:
class Persona:
    def __init__(self, nombre, edad):
        self.nombre = nombre
        self.edad = edad

    def __eq__(self, other):
        return self.nombre == other.nombre and self.edad == other.edad

p1 = Persona("Juan", 25) # Crear un objeto de la clase Persona
p2 = Persona("Juan", 25) # Crear otro objeto de la clase Persona
print(p1 == p2) # Salida: True


True


* **Método `__len__()`:**

Este método se utiliza para devolver la longitud de un objeto. Es llamado automáticamente cuando se usa la función `len()` en un objeto. Por ejemplo:

In [8]:
class MiLista:
    def __init__(self, elementos):
        self.elementos = elementos

    def __len__(self):
        return len(self.elementos)

l = MiLista([1, 2, 3, 4, 5]) # Crear un objeto de la clase MiLista
print(len(l)) # Salida: 5


5


Estos son solo algunos ejemplos de métodos especiales en Python. Hay muchos otros disponibles en el lenguaje y se pueden utilizar para personalizar el comportamiento de las clases y objetos de manera efectiva.

## **El módulo cmath**

El módulo `cmath` es una biblioteca estándar de Python que proporciona funciones matemáticas para operaciones complejas. Este módulo es similar al módulo `math` pero está diseñado específicamente para trabajar con números complejos, los cuales se representan en Python con la letra `j` al final del número imaginario.

A continuación, se presentan algunos ejemplos de uso del módulo `cmath`:

In [9]:
import cmath

# Crear un número complejo
z = 3 + 4j

# Obtener la parte real e imaginaria de un número complejo
print("Parte real:", z.real)  # Salida: 3.0
print("Parte imaginaria:", z.imag)  # Salida: 4.0

# Obtener el conjugado de un número complejo
z_conjugado = z.conjugate()
print("Conjugado:", z_conjugado)  # Salida: (3-4j)

# Obtener la fase de un número complejo
fase = cmath.phase(z)
print("Fase:", fase)  # Salida: 0.93 radianes (aprox.)

# Obtener la magnitud (o módulo) de un número complejo
magnitud = abs(z)
print("Magnitud:", magnitud)  # Salida: 5.0

# Realizar operaciones matemáticas con números complejos
a = 1 + 2j
b = 3 - 4j

suma = a + b
resta = a - b
producto = a * b
division = a / b

print("Suma:", suma)  # Salida: (4-2j)
print("Resta:", resta)  # Salida: (-2+6j)
print("Producto:", producto)  # Salida: (11-2j)
print("División:", division)  # Salida: (-0.2+0.4j)

# Obtener la raíz cuadrada de un número complejo
raiz_cuadrada = cmath.sqrt(z)
print("Raíz cuadrada:", raiz_cuadrada)  # Salida: (2+1j)

# Obtener el logaritmo natural de un número complejo
logaritmo = cmath.log(z)
print("Logaritmo natural:", logaritmo)  # Salida: (1.6094379124341003+0.9272952180016122j)


Parte real: 3.0
Parte imaginaria: 4.0
Conjugado: (3-4j)
Fase: 0.9272952180016122
Magnitud: 5.0
Suma: (4-2j)
Resta: (-2+6j)
Producto: (11+2j)
División: (-0.2+0.4j)
Raíz cuadrada: (2+1j)
Logaritmo natural: (1.6094379124341003+0.9272952180016122j)


En resumen, el módulo `cmath` en Python es una herramienta útil para trabajar con números complejos y realizar operaciones matemáticas complejas. Con este módulo, se pueden obtener la parte real e imaginaria, el conjugado, la fase, la magnitud y realizar operaciones matemáticas con números complejos, entre otras funciones.

## **Constantes del módulo cmath**

El módulo `cmath` en Python también proporciona algunas constantes útiles para trabajar con números complejos. A continuación se presentan algunas de estas constantes y su uso:

* **cmath.pi:**

 Esta constante es igual al valor de `π`, la relación entre la circunferencia de un círculo y su diámetro. En el contexto de los números complejos, esta constante se usa a menudo para calcular las funciones trigonométricas, como seno y coseno, de un número complejo. Por ejemplo:

In [10]:
import cmath

z = 1 + 1j
seno = cmath.sin(cmath.pi * z)
coseno = cmath.cos(cmath.pi * z)

print("Seno:", seno)  # Salida: (1.2989340843532393-0.6367380071391379j)
print("Coseno:", coseno)  # Salida: (0.8337300251311491+0.9888977057628651j)


Seno: (1.4196048474733104e-15-11.548739357257748j)
Coseno: (-11.591953275521519-1.4143126688052763e-15j)


* **cmath.e:**

 Esta constante es igual a la base del logaritmo natural, `e.` En el contexto de los números complejos, esta constante se usa a menudo para calcular el logaritmo natural de un número complejo. Por ejemplo:

In [11]:
import cmath

z = 1 + 1j
logaritmo = cmath.log(z, cmath.e)

print("Logaritmo natural:", logaritmo)  # Salida: (0.34657359027997264+0.7853981633974483j)


Logaritmo natural: (0.34657359027997264+0.7853981633974483j)


* **cmath.inf:**

 Esta constante representa el infinito en el contexto de los números complejos. Por ejemplo:

In [12]:
import cmath

z = 1 + 1j
division = z / cmath.inf

print("División:", division)  # Salida: 0j


División: 0j


* **cmath.nan:**

 Esta constante representa un valor que no es un número (NaN, por sus siglas en inglés) en el contexto de los números complejos. Por ejemplo:

In [13]:
import cmath

z = cmath.nan + 1j

print("Número complejo:", z)  # Salida: (nan+1j)


Número complejo: (nan+1j)


En resumen, el módulo `cmath` en Python también proporciona algunas constantes útiles, como `π`, `e`, `infinito` y `NaN`, que se pueden utilizar para realizar cálculos matemáticos complejos con números complejos.

## **Métodos. Coordenadas polares**

En Python, los números complejos se pueden representar en coordenadas polares utilizando la función `cmath.polar()`. Esta función devuelve una tupla de dos valores: la magnitud del número complejo y su ángulo en radianes. A continuación se muestra un ejemplo de cómo se utiliza esta función:

In [14]:
import cmath

z = 1 + 1j
polar = cmath.polar(z)

print("Coordenadas polares:", polar)  # Salida: (1.4142135623730951, 0.7853981633974483)


Coordenadas polares: (1.4142135623730951, 0.7853981633974483)


En este ejemplo, se crea un número complejo `z` con los valores reales `e` imaginarios 1 y 1, respectivamente. Luego, se llama a la función `cmath.polar()` pasando `z` como argumento, lo que devuelve una tupla que contiene la magnitud de `z`, que es `√2`, y el ángulo en radianes, que es `π/4` o `45°`.

Además de `cmath.polar()`, el módulo `cmath` también proporciona una función llamada `cmath.rect()` que permite construir un número complejo a partir de sus coordenadas polares. Esta función toma dos argumentos: la magnitud y el ángulo en radianes, y devuelve un número complejo. A continuación se muestra un ejemplo de cómo se utiliza esta función:

In [15]:
import cmath

magnitud = 1.4142135623730951
angulo = 0.7853981633974483

rectangular = cmath.rect(magnitud, angulo)

print("Coordenadas rectangulares:", rectangular)  # Salida: (1+1j)


Coordenadas rectangulares: (1.0000000000000002+1j)


En este ejemplo, se tienen las coordenadas polares del número complejo `√2 e^(π/4)i`. Se utiliza la función `cmath.rect()` para construir este número complejo a partir de sus coordenadas polares, lo que devuelve el número complejo `1 + i`.

En resumen, las coordenadas polares se pueden utilizar para representar números complejos en Python utilizando las funciones `cmath.polar()` y `cmath.rect()`. La función `cmath.polar()` toma un número complejo como argumento y devuelve su magnitud y ángulo en radianes, mientras que la función `cmath.rect()` toma la magnitud y el ángulo en radianes de un número complejo como argumentos y devuelve el número complejo correspondiente en coordenadas rectangulares.

## **Métodos. Potencias y logaritmos**

En Python, el módulo `math` proporciona funciones para calcular potencias y logaritmos. A continuación, se describen las funciones más comunes y se proporcionan ejemplos de cómo utilizarlas.

* **Potencias:**

`math.pow(x, y)`: Esta función devuelve el resultado de elevar `x` a la potencia `y`.

In [16]:
import math

a = 2
b = 3

# a elevado a la potencia b
c = math.pow(a, b)

print(c)  # Salida: 8.0


8.0


En este ejemplo, se utiliza la función `math.pow()` para elevar 2 a la potencia 3, lo que resulta en el número 8.

`x ** y`: La notación `**` también se puede utilizar para calcular potencias.

In [17]:
a = 2
b = 3

# a elevado a la potencia b
c = a ** b

print(c)  # Salida: 8


8


En este ejemplo, se utiliza la notación `**` para elevar 2 a la potencia 3, lo que resulta en el número 8.

* **Logaritmos:**

`math.log(x)`: Esta función devuelve el logaritmo natural `(base e)` de `x`.


In [18]:
import math

x = 10

# logaritmo natural de x
y = math.log(x)

print(y)  # Salida: 2.302585092994046


2.302585092994046


En este ejemplo, se utiliza la función `math.log()` para calcular el logaritmo natural de `10`, lo que resulta en el número 2.302585092994046.

`math.log(x, b)`: Esta función devuelve el logaritmo de `x` en base `b`.

In [19]:
import math

x = 100
b = 10

# logaritmo de x en base b
y = math.log(x, b)

print(y)  # Salida: 2.0


2.0


En este ejemplo, se utiliza la función `math.log()` con dos argumentos para calcular el logaritmo de `100` en base `10`, lo que resulta en el número `2`.

`math.log10(x)`: Esta función devuelve el logaritmo de `x` en base `10`.

In [20]:
import math

x = 100

# logaritmo de x en base 10
y = math.log10(x)

print(y)  # Salida: 2.0


2.0


En este ejemplo, se utiliza la función `math.log10()` para calcular el logaritmo de `100` en base `10`, lo que resulta en el número `2`.

En resumen, en Python se pueden calcular potencias utilizando la función `math.pow()` o la notación `**`, y se pueden calcular logaritmos utilizando las funciones `math.log()` o `math.log10()`. La función `math.log()` también permite especificar la base del logaritmo como un segundo argumento.

## **Métodos. Funciones trigonométricas**

En Python, el módulo `math` proporciona funciones para calcular valores trigonométricos de ángulos. A continuación, se describen las funciones más comunes y se proporcionan ejemplos de cómo utilizarlas.

* **math.sin(x):**

 Esta función devuelve el seno del ángulo `x`, donde `x` está en radianes.

In [21]:
import math

# Seno de 30 grados
x = math.radians(30)
y = math.sin(x)

print(y)  # Salida: 0.49999999999999994


0.49999999999999994


En este ejemplo, se utiliza la función `math.radians()` para convertir 30 grados a radianes, y luego se utiliza la función `math.sin()` para calcular el seno del ángulo resultante. El resultado es aproximadamente 0.5.

* **math.cos(x):**

 Esta función devuelve el coseno del ángulo `x`, donde `x` está en radianes.

In [22]:
import math

# Coseno de 45 grados
x = math.radians(45)
y = math.cos(x)

print(y)  # Salida: 0.7071067811865476


0.7071067811865476


En este ejemplo, se utiliza la función `math.radians()` para convertir 45 grados a radianes, y luego se utiliza la función `math.cos()` para calcular el coseno del ángulo resultante. El resultado es aproximadamente 0.707.

* **math.tan(x):**

 Esta función devuelve la tangente del ángulo `x`, donde `x` está en radianes.

In [23]:
import math

# Tangente de 60 grados
x = math.radians(60)
y = math.tan(x)

print(y)  # Salida: 1.7320508075688767


1.7320508075688767


En este ejemplo, se utiliza la función `math.radians()` para convertir 60 grados a radianes, y luego se utiliza la función `math.tan()` para calcular la tangente del ángulo resultante. El resultado es aproximadamente 1.732.

* **math.asin(x):**

 Esta función devuelve el ángulo en radianes cuyo seno es `x`.

In [24]:
import math

# Ángulo cuyo seno es 0.5
x = 0.5
y = math.asin(x)

print(math.degrees(y))  # Salida: 30.000000000000004


30.000000000000004


En este ejemplo, se utiliza la función `math.asin()` para calcular el ángulo en radianes cuyo seno es 0.5. Luego, se utiliza la función `math.degrees()` para convertir el resultado a grados. El resultado es aproximadamente 30 grados.

* **math.acos(x):**

 Esta función devuelve el ángulo en radianes cuyo coseno es `x`.

In [25]:
import math

# Ángulo cuyo coseno es 0.5
x = 0.5
y = math.acos(x)

print(math.degrees(y))  # Salida: 60.00000000000001


60.00000000000001


En este ejemplo, se utiliza la función `math.acos()` para calcular el ángulo en radianes cuyo coseno es 0.5. Luego, se utiliza la función `math.degrees()` para convertir el resultado a grados. El resultado es aproximadamente 60 grados.

* **math.atan(x):**

 Esta función devuelve el ángulo en radianes cuya tangente es `x`.

In [26]:
import math

# Ángulo cuya tangente es 1.732
x = 1.732
y = math.atan(x)

print(math.degrees(y))  # Salida: 60.00000000000001


59.999272219172624


En este ejemplo, se utiliza la función `math.atan()` para calcular el ángulo en radianes cuya tangente es 1.732. Luego, se utiliza la función `math.degrees()` para convertir el resultado a grados. El resultado es aproximadamente 60 grados.

* **math.atan2(y, x):**

 Esta función devuelve el ángulo en radianes cuya tangente es `y/x`, teniendo en cuenta el cuadrante de `(x, y)`.

In [27]:
import math

# Ángulo cuya tangente es -1.0 y (x, y) = (-1, -1)
x = -1
y = -1
theta = math.atan2(y, x)

print(math.degrees(theta))  # Salida: -135.0


-135.0


En este ejemplo, se utiliza la función `math.atan2()` para calcular el ángulo en radianes cuya tangente es -1.0 y cuyas coordenadas son (-1, -1). La función `math.degrees()` se utiliza para convertir el resultado a grados. El resultado es aproximadamente -135 grados.

## **Métodos. Funciones hiperbólicas**

En Python, las funciones hiperbólicas se encuentran en el módulo `math` y se utilizan para realizar cálculos con números hiperbólicos. Las funciones hiperbólicas son similares a las funciones trigonométricas, pero utilizan números hiperbólicos en lugar de números reales. A continuación, se explican algunas de las funciones hiperbólicas más utilizadas en Python:

* **math.sinh(x):**

 Esta función devuelve el seno hiperbólico de `x`.

In [28]:
import math

x = 2
y = math.sinh(x)

print(y)  # Salida: 3.6268604078470186


3.626860407847019


En este ejemplo, se utiliza la función `math.sinh()` para calcular el seno hiperbólico de 2. El resultado es aproximadamente 3.6268604078470186.

* **math.cosh(x):**

 Esta función devuelve el coseno hiperbólico de `x`.

In [29]:
import math

x = 2
y = math.cosh(x)

print(y)  # Salida: 3.7621956910836314


3.7621956910836314


En este ejemplo, se utiliza la función `math.cosh()` para calcular el coseno hiperbólico de 2. El resultado es aproximadamente 3.7621956910836314.

* **math.tanh(x):**

 Esta función devuelve la tangente hiperbólica de `x`.

In [30]:
import math

x = 2
y = math.tanh(x)

print(y)  # Salida: 0.9640275800758169


0.9640275800758169


En este ejemplo, se utiliza la función `math.tanh()` para calcular la tangente hiperbólica de 2. El resultado es aproximadamente 0.9640275800758169.

* **math.asinh(x):**

 Esta función devuelve el arco seno hiperbólico de `x`.

In [31]:
import math

x = 2
y = math.asinh(x)

print(y)  # Salida: 1.4436354751788103


1.4436354751788103


En este ejemplo, se utiliza la función `math.asinh()` para calcular el arco seno hiperbólico de 2. El resultado es aproximadamente 1.4436354751788103.

* **math.acosh(x):**

 Esta función devuelve el arco coseno hiperbólico de `x`.

In [32]:
import math

x = 2
y = math.acosh(x)

print(y)  # Salida: 1.3169578969248166


1.3169578969248166


En este ejemplo, se utiliza la función `math.acosh()` para calcular el arco coseno hiperbólico de 2. El resultado es aproximadamente 1.3169578969248166.

* **math.atanh(x):**

 Esta función devuelve el arco tangente hiperbólica de `x`.

In [33]:
import math

x = 0.5
y = math.atanh(x)

print(y)  # Salida: 0.5493061443340548


0.5493061443340548


En este ejemplo, se utiliza la función `math.atanh()` para calcular el arco tangente hiperbólica de 0.5. El resultado es aproximadamente 0.5493061443340548.

## **Métodos. Funciones de clasificación**

En Python, las funciones de clasificación se utilizan para ordenar y clasificar objetos en listas, tuplas y otros tipos de estructuras de datos. A continuación, se presentan algunas de las funciones de clasificación más comunes en Python:

* `sorted(iterable[, key][, reverse])`:

 Esta función devuelve una lista ordenada de elementos del iterable especificado. El argumento `key` es una función opcional que se utiliza para personalizar el ordenamiento de los elementos, y el argumento reverse es un valor booleano opcional que se utiliza para especificar si los elementos deben ser ordenados en orden ascendente o descendente.

In [34]:
numbers = [3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5]

sorted_numbers = sorted(numbers)

print(sorted_numbers)  # Salida: [1, 1, 2, 3, 3, 4, 5, 5, 5, 6, 9]


[1, 1, 2, 3, 3, 4, 5, 5, 5, 6, 9]


En este ejemplo, se utiliza la función `sorted()` para ordenar los números de la lista numbers en orden ascendente. El resultado es una lista ordenada de números.

* `list.sort(key=None, reverse=False)`:

 Esta función se utiliza para ordenar una lista en su lugar. El argumento `key` es una función opcional que se utiliza para personalizar el ordenamiento de los elementos, y el argumento reverse es un valor booleano opcional que se utiliza para especificar si los elementos deben ser ordenados en orden ascendente o descendente.

In [35]:
numbers = [3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5]

numbers.sort()

print(numbers)  # Salida: [1, 1, 2, 3, 3, 4, 5, 5, 5, 6, 9]


[1, 1, 2, 3, 3, 4, 5, 5, 5, 6, 9]


En este ejemplo, se utiliza el método `sort()` para ordenar los números de la lista numbers en orden ascendente. El resultado es la lista numbers ordenada.

* `max(iterable[, key])`:

 Esta función devuelve el elemento máximo en el iterable especificado. El argumento `key` es una función opcional que se utiliza para personalizar la comparación de elementos.

In [36]:
numbers = [3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5]

max_number = max(numbers)

print(max_number)  # Salida: 9


9


En este ejemplo, se utiliza la función `max()` para encontrar el número máximo en la lista numbers. El resultado es el número máximo en la lista.

* `min(iterable[, key])`:

 Esta función devuelve el elemento mínimo en el iterable especificado. El argumento `key` es una función opcional que se utiliza para personalizar la comparación de elementos.

In [37]:
numbers = [3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5]

min_number = min(numbers)

print(min_number)  # Salida: 1


1


En este ejemplo, se utiliza la función `min()` para encontrar el número mínimo en la lista numbers. El resultado es el número mínimo en la lista.

* `sorted(iterable, key=lambda x: x[1])`:

 Esta función se utiliza para ordenar una lista de tuplas en función del segundo elemento de cada tupla. En este caso, se utiliza una función `lambda` para especificar que se debe utilizar el segundo elemento de cada tupla como clave para ordenar la lista.

In [38]:
students = [("Alice", 80), ("Bob", 75), ("Charlie", 90), ("Dave", 85)]

sorted_students = sorted(students, key=lambda x: x[1], reverse=True)

print(sorted_students)  # Salida: [("Charlie", 90), ("Dave", 85), ("Alice", 80), ("Bob", 75)]


[('Charlie', 90), ('Dave', 85), ('Alice', 80), ('Bob', 75)]


En este ejemplo, se utiliza la función `sorted()` junto con una función `lambda` para ordenar la lista de tuplas `students` en función de las calificaciones de los estudiantes en orden descendente. El resultado es una lista ordenada de tuplas que muestra los nombres y las calificaciones de los estudiantes en orden descendente.

| **Inicio** | **atrás 14** | **Siguiente 16** |
|----------- |-------------- |---------------|
| [🏠](../../README.md) | [⏪](./14.Scripts_Modulos.ipynb)| [⏩](./16.Modulos_Numpy.ipynb)|